pubsub/kafkapubsub/kafka.go

```
@@ -407,75 +407,109 @@ func (s *subscription) Cleanup(sarama.ConsumerGroupSession) error {
-					return false
-				},
+		// We'll give up after maxWaitCtx is Done, or if s.closeCh is closed.
+		// Otherwise, we want to pull a message from one of the channels in the
+		// claim(s) we've been given.
+		//
+		// Note: we could multiplex this by ranging over each claim.Messages(),
+		// writing the messages to a single ch, and then reading from that ch
+		// here. However, this results in us reading messages from Kafka and
+		// essentially queueing them here; when the session is closed for whatever
+		// reason, those messages are lost, which may or may not be an issue
+		// depending on the Kafka configuration being used.
+		//
+		// It seems safer to use reflect.Select to explicitly only get a single
+		// message at a time, and hand it directly to the user.
+		s.mu.Lock()
+		cases := make([]reflect.SelectCase, 0, len(s.claims)+2)
+		// A case for s.closeCh being closed, at index = 0.
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(s.closeCh),
+		})
+		// A case for maxWaitCtx being Done, at index = 1.
+		cases = append(cases, reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(maxWaitCtx.Done()),
+		})
+		// A case per claim.
+		for _, claim := range s.claims {
+			cases = append(cases, reflect.SelectCase{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(claim.Messages()),
+			})
+		}
+		s.mu.Unlock()
+		i, v, ok := reflect.Select(cases)
+		if !ok {
+			switch i {
+			case 0: // s.closeCh
+				return nil, s.closeErr
+			case 1: // maxWaitCtx
+				// We've tried for a while to get maxMessages, but didn't get enough.
+				// Return whatever we have (may be empty).
+				return nil, ctx.Err()
 			}
-			dms = append(dms, dm)
-
-			s.mu.Lock()
-			s.unacked = append(s.unacked, ack)
-			s.mu.Unlock()
+			// Otherwise, if one of the claim channels closed, we're probably ending
+			// a session. Just keep trying.
+			continue
+		}
+		if v.IsNil() {
```

These next two cases should be handled somehow. If sarama's doc says or implies the channel values will never be nil, then this is a bug in their package, and you should either panic or return an Internal error.

---

OK, I removed these two cases, will just let it panic. I think it should be fine :-).